packages/ag-charts-community/src/util/panToBBox.ts

```diff
@@ -0,0 +1,64 @@
+import type { BBoxValues } from "./bboxinterface";
+import { clamp } from "./number";
+
+type Ratios = { min: number, max: number };
+type XYRatios = { x: Ratios, y: Ratios };
+
+function normalize(screenMin: number, min: number, screenMax: number, max: number, target: number): number {
+    return min + (max - min) * ((target - screenMin) / (screenMax - screenMin));
+}
+
+function unnormalize(screenMin: number, min: number, screenMax: number, max: number, ratio: number): number {
+    return screenMin + (ratio - min) * ((screenMax - screenMin) / (max - min));
+}
+function calcWorld(viewportMin: number, viewportMax: number, ratio: Ratios): [number, number] {
+    return [
+        unnormalize(viewportMin, ratio.min, viewportMax, ratio.max, 0),
+        unnormalize(viewportMin, ratio.min, viewportMax, ratio.max, 1),
+    ];
+}
+/* Pan viewport min (unnormalised, i.e. pixel coords.) by the smallest amount
+   such that the viewport range includes the input target range but clamped at
+   the world range. The function assumes:
+   1)  worldMin <= viewportMin <= viewportMax <= worldMax
+   2)  (viewportMax - viewportMin) >= (targetMax - targetMin)
+*/
+function panAxesUnnormalized(
+    worldMin: number,
+    worldMax: number,
+    viewportMin: number,
+    viewportMax: number,
+    targetMin: number,
+    targetMax: number
+): number {
+    if (viewportMin <= targetMin && targetMax <= viewportMax) return viewportMin;
+    const mindiff = targetMin - viewportMin;
+    const maxdiff = targetMax - viewportMax;
+    const diff = Math.abs(mindiff) < Math.abs(maxdiff) ? mindiff : maxdiff;
+    return clamp(worldMin, viewportMin + diff, worldMax);
+}
+
+// The calculations of the new desired viewport (i.e. ZoomState) is done in pixel coords (unnormalised).
+// The desired (x, y) for the new viewport is found, the pixel coords are converted into normalized values
+export function calcPanToBBoxRatios(viewport: BBoxValues, ratioX: Ratios, ratioY: Ratios, target: BBoxValues): XYRatios {
+    const [targetXMin, targetXMax] = [target.x, target.x + target.width];
+    const [targetYMin, targetYMax] = [target.y, target.y + target.height];
+    const [viewportXMin, viewportXMax] = [viewport.x, viewport.x + viewport.width];
+    const [viewportYMin, viewportYMax] = [viewport.y, viewport.y + viewport.height];
```

Comment:
Entirely optional, but I added a `Vec4.from(bbox) === { x1: x, y1: y, x2: x + width, y2: y + height }` util in `vector4.ts` as I was doing lots of work with bboxes and coordinates.

It might help with some of the repetition (and similarly in the `containsBBox()` function perhaps).

```suggestion
    const target = Vec4.from(target);
    const viewport = Vec4.from(viewport);
```

Reply:
Good suggestion 👍  `Vector4` feels a bit like a misnomer. Perhaps `Line2D` is a better name. But that's a discussion for another time.